presto-orc/src/main/java/com/facebook/presto/orc/writer/StreamOrderingLayout.java

```diff
@@ -0,0 +1,212 @@
+        // <Column Id, KeyId> -> SequenceId
+        private final Map<ColumnKeyInfo, Integer> keyToSequence;
+
+        public StreamMetadata(Map<ColumnSequenceInfo, List<StreamDataOutput>> sequenceToStreams, Map<ColumnKeyInfo, Integer> keyToSequence)
+        {
+            this.sequenceToStreams = requireNonNull(sequenceToStreams, "sequenceToStreams cannot be null");
+            this.keyToSequence = requireNonNull(keyToSequence, "keyToSequence cannot be null");
+        }
+    }
+
+    private static class ColumnKeyInfo
+    {
+        private final int column;
+        private final DwrfProto.KeyInfo key;
+
+        public ColumnKeyInfo(int column, DwrfProto.KeyInfo key)
+        {
+            this.column = column;
+            this.key = requireNonNull(key, "inputKeyId cannot be null");
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (obj == null) {
+                return false;
+            }
+            if (!(obj instanceof ColumnKeyInfo)) {
+                return false;
+            }
+            ColumnKeyInfo input = (ColumnKeyInfo) obj;
+            return this.column == input.column && this.key.equals(input.key);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(column, key);
+        }
+    }
+
+    private static class ColumnSequenceInfo
+    {
+        private final int column;
+        private final int sequence;
+
+        public ColumnSequenceInfo(int column, int sequence)
+        {
+            this.column = column;
+            this.sequence = sequence;
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (obj == null) {
+                return false;
+            }
+            if (!(obj instanceof ColumnSequenceInfo)) {
+                return false;
+            }
+            ColumnSequenceInfo input = (ColumnSequenceInfo) obj;
+            return this.column == input.column && this.sequence == input.sequence;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(column, sequence);
+        }
+    }
+
+    private StreamMetadata getStreamMetadata(
+            Map<Integer, Integer> nodeIdToColumn,
+            Map<Integer, ColumnEncoding> nodeIdToColumnEncodings,
+            DwrfStreamOrderingConfig config)
+    {
+        ImmutableMap.Builder<ColumnKeyInfo, Integer> keyToSequenceBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<ColumnSequenceInfo, List<StreamDataOutput>> sequenceToStreamsBuilder = ImmutableMap.builder();
+        Map<Integer, Set<DwrfProto.KeyInfo>> columnToKeySet = config.getColumnToKeySet();
+
+        // iterate through all the encodings and if the encoding has additional sequence encodings put it in the map
+        for (Map.Entry<Integer, ColumnEncoding> entry : nodeIdToColumnEncodings.entrySet()) {
+            int nodeId = entry.getKey();
```

Comment:
The value column can be a nested column, in which case multiple node ids will belong to the same column. Since StreamMetadata return is based on the column, wouldn't this method keep on overwriting the data. Also if some child nodes have some sequence missing (can it happen in list/map/struct case), there might be some streams that might get moved towards the end, instead of grouping together.

Reply:
> The value column can be a nested column, in which case multiple node ids will belong to the same column. Since StreamMetadata return is based on the column, wouldn't this method keep on overwriting the data.

I am not following this, may be explain this with an example?

To provide some context, there will only be one columnEncoding for a given nodeId and this will be fetched from [here](https://github.com/harsharastogi/presto/blob/master/presto-orc/src/main/java/com/facebook/presto/orc/writer/StructColumnWriter.java#L101) in the main writer. 

In the case of flat map column the column encoding for the root map will contain additional sequence encodings for each of the sequence. Also, the method `createColumnEncodings` in the test case shows how this map is expected to be for a given file schema

> Also if some child nodes have some sequence missing (can it happen in list/map/struct case), there might be some streams that might get moved towards the end, instead of grouping together.

I do not know follow this. Can you explain through an example?